samples/drivers/CAN/src/main.c

```diff
@@ -208,73 +136,52 @@ void main(void)
+	led_gpio_dev = device_get_binding(DT_ALIAS_LED0_GPIOS_CONTROLLER);
 	if (!led_gpio_dev) {
 		printk("LED: Device driver not found.\n");
 		return;
 	}
 
-	k_sem_init(&tx_sem, 0, INT_MAX);
-
-	button_gpio_dev = device_get_binding(CONFIG_GPIO_BUTTON_DEV);
-	if (!button_gpio_dev) {
-		printk("Button: Device driver not found.\n");
-		return;
-	}
-
-	ret = gpio_pin_configure(button_gpio_dev, CONFIG_PIN_USER_BUTTON,
-				    (GPIO_DIR_IN | GPIO_INT | GPIO_INT_EDGE |
-				     GPIO_INT_ACTIVE_HIGH | GPIO_INT_DEBOUNCE));
-	if (ret) {
-		printk("Error configuring  button pin\n");
-	}
-
-	gpio_init_callback(&gpio_cb, button_callback,
-			   BIT(CONFIG_PIN_USER_BUTTON));
-
-	ret = gpio_add_callback(button_gpio_dev, &gpio_cb);
-	if (ret) {
-		printk("Cannot setup callback!\n");
-	}
-
-	ret = gpio_pin_enable_callback(button_gpio_dev, CONFIG_PIN_USER_BUTTON);
-	if (ret) {
-		printk("Error enabling callback!\n");
+	ret = gpio_pin_configure(led_gpio_dev, DT_ALIAS_LED0_GPIOS_PIN,
+				 GPIO_DIR_OUT);
+	if (ret < 0) {
+		printk("Error setting LED pin to output mode [%d]", ret);
 	}
+#endif
 
-	ret = can_attach_isr(can_dev, rx_button_isr, 0, &filter);
+	ret = can_attach_workq(can_dev, &k_sys_work_q, &rx_work, change_led,
+			       led_gpio_dev, &change_led_filter);
 	if (ret == CAN_NO_FREE_FILTER) {
 		printk("Error, no filter available!\n");
 		return;
 	}
 
-	k_tid_t tx_tid = k_thread_create(&tx_thread_data, tx_thread_stack,
-					 K_THREAD_STACK_SIZEOF(tx_thread_stack),
-					 tx_thread,
+	printk("Change LED filter ID: %d\n", ret);
+
+	k_tid_t rx_tid = k_thread_create(&rx_thread_data, rx_thread_stack,
+					 K_THREAD_STACK_SIZEOF(rx_thread_stack),
+					 rx_thread,
 					 can_dev, NULL, NULL,
-					 TX_THREAD_PRIORITY, 0, K_NO_WAIT);
-	if (!tx_tid) {
-		printk("ERROR spawning tx_thread\n");
+					 RX_THREAD_PRIORITY, 0, K_NO_WAIT);
+	if (!rx_tid) {
+		printk("ERROR spawning rx thread\n");
 	}
 
-	k_tid_t led_tid = k_thread_create(&led_thread_data, led_thread_stack,
-					  K_THREAD_STACK_SIZEOF(led_thread_stack),
-					  led_thread,
-					  &led_msgq, can_dev, led_gpio_dev,
-					  TX_THREAD_PRIORITY, 0, K_NO_WAIT);
-	if (!led_tid) {
-		printk("ERROR spawning led_thread\n");
-	}
+	printk("Finished init.\n");
 
-	k_tid_t str_tid = k_thread_create(&rx_str_thread_data,
-					  rx_str_thread_stack,
-					  K_THREAD_STACK_SIZEOF(rx_str_thread_stack),
-					  rx_str_thread,
-					  &str_msgq, can_dev, NULL,
-					  TX_THREAD_PRIORITY, 0, K_NO_WAIT);
-	if (!str_tid) {
-		printk("ERROR spawning str_thread\n");
+	while (1) {
+		change_led_frame.data[0] = toggle++ & 0x01 ? SET_LED : RESET_LED;
+		/* This sending call is none blocking. */
+		can_send(can_dev, &change_led_frame, K_FOREVER, tx_irq_callback,
+			 "LED change");
+		k_sleep(SLEEP_TIME);
+
+		UNALIGNED_PUT(sys_cpu_to_be16(counter),
+			      (u16_t *)&counter_frame.data[0]);
+		counter++;
+		/* This sending call is blocking until the message is sent. */
```

Comment:
Technically this is blocking until timeout or sent, could probably just live without the comments, the intention throughout the sample is quite clear to cover two tx paths (blocking/non blocking) and two rx paths (work queue and polling on msgq) :-)

Reply:
@karstenkoenig the timeout is just for the mailbox acquisition. If the bus is busy, it waits "forever".

Reply:
Hmm right, yeah I remember I just followed the stm32 can driver there and try acquiring a tx buffer slot with a timeout. The blocking nature of the call comes from not providing a tx finished callback. 

Not sure about changing the behavior, right now there is no abort tx (except reinit the driver) and I'd just wait until the need/requirement comes up to have a use case to validate :-) I haven't seen a need yet.